src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1554,84 +1440,254 @@ ExtractInsertPartitionValue(Query *query, Var *partitionColumn)
+ * table by looking into range table entries.
+ */
+static bool
+SelectsFromDistributedTable(List *rangeTableList)
+{
+	ListCell *rangeTableCell = NULL;
+	foreach(rangeTableCell, rangeTableList)
+	{
+		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(rangeTableCell);
+		DistTableCacheEntry *cacheEntry = NULL;
+
+		if (rangeTableEntry->relid == InvalidOid)
+		{
+			continue;
+		}
+
+		cacheEntry = DistributedTableCacheEntry(rangeTableEntry->relid);
+		if (cacheEntry->partitionMethod != DISTRIBUTE_BY_NONE &&
+			!UpdateOrDeleteRTE(rangeTableEntry))
+		{
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * RouterQuery runs router pruning logic for SELECT, UPDATE and DELETE queries.
+ * If there are shards present and query is routable, all RTEs have been updated
+ * to point to the relevant shards in the originalQuery. Also, placementList is
+ * filled with the list of worker nodes that has all the required shard placements
+ * for the query execution. anchorShardId is set to the first pruned shardId of
+ * the given query. Finally, relationShardList is filled with the list of
+ * relation-to-shard mappings for  the query.
  *
- * On return true, all RTEs have been updated to point to the relevant shards in
- * the originalQuery. Also, placementList is filled with the list of worker nodes
- * that has all the required shard placements for the query execution.
- * anchorShardId is set to the first pruned shardId of the given query. Finally,
- * relationShardList is filled with the list of relation-to-shard mappings for
- * the query.
+ * If the given query is not routable, it fills planningError with the related
+ * DeferredErrorMessage. The caller can check this error message to see if query
+ * is routable or not.
  */
-bool
-RouterSelectQuery(Query *originalQuery, RelationRestrictionContext *restrictionContext,
-				  List **placementList, uint64 *anchorShardId, List **relationShardList,
-				  bool replacePrunedQueryWithDummy)
+DeferredErrorMessage *
+PlanRouterQuery(Query *originalQuery, RelationRestrictionContext *restrictionContext,
+				List **placementList, uint64 *anchorShardId, List **relationShardList,
+				bool replacePrunedQueryWithDummy)
 {
-	List *prunedRelationShardList = TargetShardIntervalsForSelect(originalQuery,
+	List *prunedRelationShardList = TargetShardIntervalsForRouter(originalQuery,
 																  restrictionContext);
-	uint64 shardId = INVALID_SHARD_ID;
 	CmdType commandType PG_USED_FOR_ASSERTS_ONLY = originalQuery->commandType;
+	DeferredErrorMessage *planningError = NULL;
 	ListCell *prunedRelationShardListCell = NULL;
 	List *workerList = NIL;
 	bool shardsPresent = false;
+	uint64 shardId = INVALID_SHARD_ID;
 
 	*placementList = NIL;
 
+	/*
+	 * If prunedRelationShardList is NULL then it means a relation has more
+	 * than one shard left after pruning.
+	 *
+	 * TODO: The way TargetShardIntervalsForRouter represents multiple shards
```

Comment:
I added a commit to fix this. I think I can also squash that one given that I also modified other parts of this function.